src/dataframe/dataframe.jl

```diff
@@ -846,16 +847,216 @@ function _deleteat!_helper(df::DataFrame, drop)
     return df
 end
 
+"""
+    keepat!(df::DataFrame, inds)
+
+Delete rows at all indices not specified by `inds` from a `DataFrame` `df`
+in place and return it.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     2      5
+   3 │     3      6
+
+julia> keepat!(df, [1, 3])
+2×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     3      6
+```
+"""
+function keepat!(df::DataFrame, inds)
+    if !issorted(inds, lt=<=)
+        throw(ArgumentError("Indices passed to keepat! must be unique and sorted"))
+    end
+    return deleteat!(df, Not(inds))
+end
+
+function keepat!(df::DataFrame, inds::Integer)
+    inds isa Bool && throw(ArgumentError("Invalid index of type Bool"))
+    return keepat!(df, Int[inds])
+end
+
+keepat!(df::DataFrame, inds::AbstractVector{Bool}) = deleteat!(df, .!inds)
+keepat!(df::DataFrame, inds::Not) = keepat!(df, axes(df, 1)[inds])
+
 """
     empty!(df::DataFrame)
 
 Remove all rows from `df`, making each of its columns empty.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     2      5
+   3 │     3      6
+
+julia> empty!(df)
+0×2 DataFrame
+
+julia> Tables.columntable(df)
+(a = Int64[], b = Int64[])
+```
+
 """
 function Base.empty!(df::DataFrame)
     foreach(empty!, eachcol(df))
     return df
 end
 
+"""
+    resize!(df::DataFrame, n::Integer)
+
+Resize `df` to have `n` rows by calling `resize!` on all columns of `df`.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     2      5
+   3 │     3      6
+
+julia> resize!(df, 2)
+2×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     2      5
+```
+"""
+function Base.resize!(df::DataFrame, n::Integer)
+    if ncol(df) == 0 && n != 0
+        throw(ArgumentError("data frame has no columns and requested number " *
+                            "of rows is not zero"))
+    end
+    foreach(col -> resize!(col, n), eachcol(df))
+    return df
+end
+
+"""
+    pop!(df::DataFrame)
+
+Remove the last row from `df` and return a `NamedTuple` created from this row.
+
+!!! note
+
+    Using this method for very wide data frames may lead to expensive compilation.
```

Comment:
AFAICT this is the only place in the API where we return a `NamedTuple`, right? I agree that's probably the best choice, but let's confirm this. The only alternative I can see would be to create a new `CopiedDataFrameRow` type that would wrap a `Vector{Any}` of values (and could still point to `df`'s column index). That would likely be quite slow. Right?

Reply:
The alternative would be to return `first(df[i:i, :])` - i.e. `DataFrameRow` but on a copied 1-row DataFrame.

I was also considering a kwarg allowing to choose between `NamedTuple` and `DataFrameRow`.

Here is a comparison for a narrow case (as you can see the speed is comparable)
```
julia> df = DataFrame(a=1:2, b=1.0:2.0, c='a':'b')
2×3 DataFrame
 Row │ a      b        c
     │ Int64  Float64  Char
─────┼──────────────────────
   1 │     1      1.0  a
   2 │     2      2.0  b

julia> @time first(df[2:2, :])
  0.000078 seconds (17 allocations: 1.125 KiB)
DataFrameRow
 Row │ a      b        c
     │ Int64  Float64  Char
─────┼──────────────────────
   1 │     2      2.0  b

julia> @time NamedTuple(df[2, :])
  0.000029 seconds (7 allocations: 288 bytes)
(a = 2, b = 2.0, c = 'b')

julia> @benchmark first($df[2:2, :])
BenchmarkTools.Trial: 10000 samples with 126 evaluations.
 Range (min … max):  726.984 ns … 32.334 μs  ┊ GC (min … max): 0.00% … 96.93%
 Time  (median):     807.143 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   905.540 ns ±  1.335 μs  ┊ GC (mean ± σ):  7.87% ±  5.18%

     ▁▃█▆▁
  ▃▄██████▆▄▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  727 ns          Histogram: frequency by time         1.57 μs <

 Memory estimate: 1.03 KiB, allocs estimate: 15.

julia> @benchmark NamedTuple($df[2, :])
BenchmarkTools.Trial: 10000 samples with 69 evaluations.
 Range (min … max):  840.580 ns …  42.955 μs  ┊ GC (min … max): 0.00% … 96.85%
 Time  (median):     943.478 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   968.572 ns ± 712.572 ns  ┊ GC (mean ± σ):  1.25% ±  1.68%

        ▄ █ ▇ ▄
  ▆▂█▃▆▃█▅█▆█▇█▇▄▅▂▂▁▁▁▁▁▁▁▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  841 ns           Histogram: frequency by time         1.47 μs <

 Memory estimate: 224 bytes, allocs estimate: 6.
```

For 100 column data frame it is also comparable:
```
julia> df = DataFrame(rand(10, 100), :auto);

julia> @benchmark first($df[2:2, :])
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  15.700 μs …  3.628 ms  ┊ GC (min … max): 0.00% … 99.39%
 Time  (median):     17.500 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.676 μs ± 60.455 μs  ┊ GC (mean ± σ):  5.28% ±  1.72%

   ▁█▇
  ▅███▆▄▅▃▃▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂ ▃
  15.7 μs         Histogram: frequency by time          51 μs <

 Memory estimate: 15.86 KiB, allocs estimate: 209.

julia> @benchmark NamedTuple($df[2, :])
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  14.800 μs … 681.800 μs  ┊ GC (min … max): 0.00% … 96.66%
 Time  (median):     16.700 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   17.465 μs ±   9.688 μs  ┊ GC (mean ± σ):  0.74% ±  1.36%

    ▃▃▂█▂ ▁▁▃
  ▅▅█████▅███▅▅▅▆▄▄▄▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▁▂▁▂▂▁▁▁▂▂▁▂▁▁▂▂▂▁▂▂▂▂▂▂▂▂ ▃
  14.8 μs         Histogram: frequency by time         31.2 μs <

 Memory estimate: 9.09 KiB, allocs estimate: 211.
```

so - what do you think? 😄

Reply:
For a reference - the other place where we return a `NamedTuple` is `copy` of a `DataFrameRow`:
```
julia> df = DataFrame(rand(10, 4), :auto)
10×4 DataFrame
 Row │ x1         x2          x3          x4
     │ Float64    Float64     Float64     Float64
─────┼──────────────────────────────────────────────
   1 │ 0.266049   0.536285    0.497338    0.946282
   2 │ 0.934083   0.113396    0.730271    0.641962
   3 │ 0.505046   0.418804    0.955571    0.228884
   4 │ 0.339139   0.867827    0.191675    0.413833
   5 │ 0.953402   0.790694    0.0697463   0.508785
   6 │ 0.432954   0.288875    0.525401    0.164282
   7 │ 0.0115583  0.803558    0.798115    0.205552
   8 │ 0.124769   0.196324    0.733265    0.674536
   9 │ 0.471618   0.00161109  0.00377343  0.923914
  10 │ 0.0981889  0.437943    0.293618    0.0527693

julia> df[1, :]
DataFrameRow
 Row │ x1        x2        x3        x4
     │ Float64   Float64   Float64   Float64
─────┼────────────────────────────────────────
   1 │ 0.266049  0.536285  0.497338  0.946282

julia> copy(df[1, :])
(x1 = 0.2660489743436787, x2 = 0.5362852297280495, x3 = 0.4973378729931356, x4 = 0.9462823778051903)
```

Reply:
I'm surprised that allocating a new data frame is as fast as creating a named tuple, especially when the number of columns is small. Anyway returning a `DataFrameRow` whose parent isn't the original data frame (as it's not possible) sounds weird...

Reply:
OK - so I understand we stay with `NamedTuple` consistently with `copy`.